Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+/* Copyright 2022 David Grote
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "WarpX.H"
+
+#include "BoundaryConditions/PML.H"
+#include "Diagnostics/MultiDiagnostics.H"
+#include "Diagnostics/ReducedDiags/MultiReducedDiags.H"
+#include "Evolve/WarpXDtType.H"
+#include "Evolve/WarpXPushType.H"
+#include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceSolver.H"
+#ifdef WARPX_USE_PSATD
+#   ifdef WARPX_DIM_RZ
+#       include "FieldSolver/SpectralSolver/SpectralSolverRZ.H"
+#   else
+#       include "FieldSolver/SpectralSolver/SpectralSolver.H"
+#   endif
+#endif
+#include "Parallelization/GuardCellManager.H"
+#include "Particles/MultiParticleContainer.H"
+#include "Particles/ParticleBoundaryBuffer.H"
+#include "Python/callbacks.H"
+#include "Utils/TextMsg.H"
+#include "Utils/WarpXAlgorithmSelection.H"
+#include "Utils/WarpXUtil.H"
+#include "Utils/WarpXConst.H"
+#include "Utils/WarpXProfilerWrapper.H"
+
+#include <ablastr/utils/SignalHandling.H>
+#include <ablastr/warn_manager/WarnManager.H>
+
+#include <AMReX.H>
+#include <AMReX_Array.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_Geometry.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LayoutData.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Print.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Utility.H>
+#include <AMReX_Vector.H>
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <ostream>
+#include <vector>
+
+void
+WarpX::ImplicitPreRHSOp ( amrex::Real  a_cur_time,
+                          amrex::Real  a_full_dt,
+                          int          a_nl_iter,
+                          bool         a_from_jacobian )
+{
+    using namespace amrex::literals;
+    amrex::ignore_unused( a_full_dt, a_nl_iter, a_from_jacobian );
+
+    // Advance the particle positions by 1/2 dt,
+    // particle velocities by dt, then take average of old and new v,
+    // deposit currents, giving J at n+1/2
+    // This uses Efield_fp and Bfield_fp, the field at n+1/2 from the previous iteration.
+    const bool skip_current = false;
+    const PushType push_type = PushType::Implicit;
+    PushParticlesandDeposit(a_cur_time, skip_current, push_type);
```

Comment:
This eventually calls [`WarpXParticleContainer::Evolve`](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/WarpXParticleContainer.H#L148), which is a virtual function and can call one of these two functions:
- [`LaserParticleContainer::Evolve`](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/LaserParticleContainer.cpp#L557)
- [`PhysicalParticleContainer::Evolve`](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/PhysicalParticleContainer.cpp#L2007)

Reply:
[PhysicalParticleContainer::Evolve](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/PhysicalParticleContainer.cpp#L2007) has been updated to correctly take into account `push_type`, but [LaserParticleContainer::Evolve](https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/LaserParticleContainer.cpp#L557) has not.